src/v/kafka/server/group_manager.cc

```diff
@@ -329,24 +338,27 @@ ss::future<> group_manager::recover_partition(
 
 ss::future<ss::stop_iteration>
 recovery_batch_consumer::operator()(model::record_batch batch) {
-    if (unlikely(batch.header().type != raft::data_batch_type)) {
-        klog.trace("ignorning batch with type {}", int(batch.header().type));
-        return ss::make_ready_future<ss::stop_iteration>(
-          ss::stop_iteration::no);
-    }
     if (as.abort_requested()) {
         return ss::make_ready_future<ss::stop_iteration>(
           ss::stop_iteration::yes);
     }
-    batch_base_offset = batch.base_offset();
-    return ss::do_with(
-             std::move(batch),
-             [this](model::record_batch& batch) {
-                 return model::for_each_record(batch, [this](model::record& r) {
-                     return handle_record(std::move(r));
-                 });
-             })
-      .then([] { return ss::stop_iteration::no; });
+
+    if (batch.header().type == raft::data_batch_type) {
+        batch_base_offset = batch.base_offset();
+        return ss::do_with(
+                 std::move(batch),
+                 [this](model::record_batch& batch) {
+                     return model::for_each_record(
```

Comment:
this is wild. the model::for_each_record takes a const ref but somehow gets a model::record-ref! because it materializes the records lazily as a copy.

Reply:
should really read, `copy-each-record`

Reply:
That's what we do even without this PR - https://github.com/vectorizedio/redpanda/blob/33669c6596080af3d44e77bb3f4ca19d6df66da4/src/v/kafka/server/group_manager.cc#L381 Changing it is out of scope of this PR, do you want create a issue?